dask/highlevelgraph.py

```diff
@@ -959,11 +959,19 @@ def cull(self, keys: Iterable) -> HighLevelGraph:
 
                 # Save the culled layer and its key dependencies
                 ret_layers[layer_name] = culled_layer
-                if isinstance(layer, Blockwise) or (
-                    layer.is_materialized() and (len(layer) == len(culled_deps))
+                if (
+                    isinstance(layer, Blockwise)
+                    or isinstance(layer, MaterializedLayer)
+                    or (layer.is_materialized() and (len(layer) == len(culled_deps)))
```

Comment:
Why was this check not already catching `MaterializedLayer`s?

If I comment out the additional `isinstance(layer, MaterializedLayet)`, then the new test still passes for me.

Reply:
Sorry - I'm just realizing that this test doesn't really cover the change in this PR. Rather, the test essentially acts as an alarm in case some future change to `MaterializedLayer` results in the `cull` method returning true "external" dependencies.

I'm having trouble thinking of a good way to design a test that specifically targets the change in this PR. The problem is that there wasn't anything "wrong" about the code before. We were just missing an optimization with significant performance repercussions.  I'm very open to suggestions here :)

Reply:
Shouldn't 
```python
 (layer.is_materialized() and (len(layer) == len(culled_deps))) 
```
already be satisfied for `MaterializedLayer` instances?

Reply:
No. If the number of requested keys is a **proper** subset of all output keys of `layer`, then `len(culled_deps)` will be less than `len(layer)`. This is why the zarr workflow was not caching.

Reply:
I see, thanks for the explanation

Reply:
Your [example](https://github.com/dask/dask/pull/9413#issuecomment-1225013372) test reminded me of `key_dependencies` (sorry for missing that earlier). I updated the test so that it will fail without the change in this PR. Thanks for your help with this!